stdlib/ballerina-database/src/main/java/org/ballerinalang/database/sql/SQLDataIterator.java

```diff
@@ -325,6 +311,261 @@ private BStruct createUserDefinedType(Struct structValue, BStructType structType
+        double dValue = rs.getDouble(index);
+        handleDoubleValue(dValue, bStruct, refRegIndex, doubleRegIndex, index);
+    }
+
+    private void handleFloatValue(BStruct bStruct, RegistryIndex refRegIndex, RegistryIndex doubleRegIndex, int index)
+            throws SQLException {
+        double fValue = rs.getFloat(index);
+        handleDoubleValue(fValue, bStruct, refRegIndex, doubleRegIndex, index);
+    }
+
+    private void handleDecimalValue(BStruct bStruct, RegistryIndex refRegIndex, RegistryIndex doubleRegIndex, int index)
+            throws SQLException {
+        double decimalValue = 0;
+        BigDecimal bigDecimalValue = rs.getBigDecimal(index);
+        if (bigDecimalValue != null) {
+            decimalValue = bigDecimalValue.doubleValue();
+        }
+        handleDoubleValue(decimalValue, bStruct, refRegIndex, doubleRegIndex, index);
+    }
+
+    private void handleDoubleValue(double fValue, BStruct bStruct, RegistryIndex refRegIndex,
+            RegistryIndex doubleRegIndex, int index) throws SQLException {
+        BStructType.StructField[] structFields = getStructFields();
+        boolean isOriginalValueNull = rs.wasNull();
+        int fieldTypeTag = getFieldTypeTag(structFields, index);
+
+        if (fieldTypeTag == TypeTags.UNION_TAG) {
+            BRefType refValue = isOriginalValueNull ? null : new BFloat(fValue);
+            validateAndSetRefRecordField(bStruct, refRegIndex.incrementAndGet(), TypeTags.FLOAT_TAG,
+                    retrieveType(structFields, index - 1), refValue);
+        } else {
+            if (isOriginalValueNull) {
+                handleNilToNonNillableFieldAssignment();
+            } else {
+                bStruct.setFloatField(doubleRegIndex.incrementAndGet(), fValue);
+            }
+        }
+    }
+
+    private void handleMappingDateValueToUnionType(BStructType.StructField[] structFields, int index, BStruct bStruct,
+            RegistryIndex refRegIndex, java.util.Date date) {
+        int type = retrieveType(structFields, index - 1);
+        switch (type) {
+        case TypeTags.STRING_TAG:
+            String dateValue = SQLDatasourceUtils.getString(date);
+            bStruct.setRefField(refRegIndex.incrementAndGet(), dateValue != null ? new BString(dateValue) : null);
+            break;
+        case TypeTags.STRUCT_TAG:
+            bStruct.setRefField(refRegIndex.incrementAndGet(), date != null ? createTimeStruct(date.getTime()) : null);
+            break;
+        case TypeTags.INT_TAG:
+            bStruct.setRefField(refRegIndex.incrementAndGet(), date != null ? new BInteger(date.getTime()) : null);
+            break;
+        default:
+            handleMismatchingFieldAssignment();
+        }
+    }
+
+    private void handleMappingDateValueToNonUnionType(java.util.Date date, int fieldTypeTag, BStruct bStruct,
+            RegistryIndex stringRegIndex, RegistryIndex refRegIndex, RegistryIndex longRegIndex) {
+        if (date != null) {
```

Comment:
Don't we need to validate using validateAndSetRefRecordField method?

Reply:
For Date, we have several possible assignable field types "string", "struct" and "int". So there is a switch below this line of which the default case is throwing an exception. So the validation happens. 

validateAndSetRefRecordField is used for other types which have only one field type.